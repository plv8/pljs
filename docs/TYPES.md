# PLJS Types

In PLJS, types are converted between Postgres native types and JavaScript types. This is not always a 1:1 mapping, but where possible there are direct conversions:

| PostgreSQL Type  | JavaScript Type      |
| ---------------- | -------------------- |
| `TEXT`           | `String`             |
| `VARCHAR`        | `String`             |
| `FLOAT4`         | `Number`             |
| `FLOAT8`         | `Number`             |
| `NUMERIC`        | `Number` or `BigInt` |
| `OID`            | `Number`             |
| `BOOL`           | `Bool`               |
| `INT2`           | `Number`             |
| `INT4`           | `Number` or `BigInt` |
| `INT8`           | `Number` or `BigInt` |
| `FLOAT4`         | `Number`             |
| `FLOAT8`         | `Number`             |
| `JSON`           | `JSON`               |
| `JSONB`          | `JSON`               |
| `TEXT`           | `String`             |
| `VARCHAR`        | `String`             |
| `BPCHAR`         | `String`             |
| `NAME`           | `String`             |
| `XML`            | `String`             |
| `BYTEA`          | `Array`              |
| `DATE`           | `Date`               |
| `TIMESTAMP`      | `Date`               |
| `TIMESTAMPTZOID` | `Date`               |
| default          | `String` or `Number` |
