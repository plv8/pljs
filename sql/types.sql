CREATE FUNCTION return_void() RETURNS void AS $$ $$ LANGUAGE pljs;
SELECT return_void();

CREATE FUNCTION return_null() RETURNS text AS $$ return null; $$ LANGUAGE pljs;
SELECT r, r IS NULL AS isnull FROM return_null() AS r;

-- TYPE CONVERTIONS
CREATE FUNCTION int2_to_int4(x int2) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT int2_to_int4(24::int2);
CREATE FUNCTION int4_to_int2(x int4) RETURNS int2 AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_int2(42);
CREATE FUNCTION int4_to_int8(x int4) RETURNS int8 AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_int8(48);
CREATE FUNCTION int8_to_int4(x int8) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT int8_to_int4(84);
CREATE FUNCTION float8_to_numeric(x float8) RETURNS numeric AS $$ return x; $$ LANGUAGE pljs;
SELECT float8_to_numeric(1.5);
CREATE FUNCTION numeric_to_int8(x numeric) RETURNS int8 AS $$ return x; $$ LANGUAGE pljs;
SELECT numeric_to_int8(1234.56);
CREATE FUNCTION int4_to_text(x int4) RETURNS text AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_text(123);
CREATE FUNCTION text_to_int4(x text) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT text_to_int4('123');
SELECT text_to_int4('abc');

-- ARRAYS
CREATE FUNCTION return_array() RETURNS TEXT[] AS $$ return ["foo", "bar"]; $$LANGUAGE pljs;
SELECT return_array();
