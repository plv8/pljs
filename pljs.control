comment = 'PL/JS trusted procedural language'

default_version = '0.1.0'
module_pathname = '$libdir/pljs'
relocatable = false
schema = pg_catalog
superuser = true
