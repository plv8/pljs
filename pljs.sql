CREATE OR REPLACE FUNCTION pljs_call_handler() RETURNS language_handler
AS 'MODULE_PATHNAME' LANGUAGE C;

CREATE OR REPLACE FUNCTION pljs_call_validator(oid) RETURNS void
	AS 'MODULE_PATHNAME' LANGUAGE C;

CREATE OR REPLACE FUNCTION pljs_inline_handler(internal) RETURNS void
  AS 'MODULE_PATHNAME' LANGUAGE C;

CREATE OR REPLACE TRUSTED LANGUAGE pljs
 HANDLER pljs_call_handler
 INLINE pljs_inline_handler
 VALIDATOR pljs_call_validator;

CREATE OR REPLACE FUNCTION pljs_version()
 RETURNS TEXT
 AS 'MODULE_PATHNAME', 'pljs_version'
 LANGUAGE C STRICT IMMUTABLE;
