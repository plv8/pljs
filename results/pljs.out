-- CREATE FUNCTION
CREATE FUNCTION pljs_test(keys text[], vals text[]) RETURNS text AS
$$
	var o = {};
	for (var i = 0; i < keys.length; i++)
		o[keys[i]] = vals[i];
	return JSON.stringify(o);
$$
LANGUAGE pljs IMMUTABLE STRICT;
SELECT pljs_test(ARRAY['name', 'age'], ARRAY['Tom', '29']);
NOTICE:  Unknown type: 1009
NOTICE:  Unknown type: 1009
ERROR:  execution error
DETAIL:  TypeError: cannot read property 'length' of null
    at pljs_test (<function>:4)

CREATE FUNCTION unnamed_args(text[], text[]) RETURNS text[] AS
$$
	var array1 = arguments[0];
	var array2 = $2;
	return array1.concat(array2);
$$
LANGUAGE pljs IMMUTABLE STRICT;
SELECT unnamed_args(ARRAY['A', 'B'], ARRAY['C', 'D']);
NOTICE:  Unknown type: 1009
NOTICE:  Unknown type: 1009
ERROR:  execution error
DETAIL:  TypeError: cannot read property 'concat' of null
    at unnamed_args (<function>:5)

CREATE FUNCTION concat_strings(VARIADIC args text[]) RETURNS text AS
$$
	var result = "";
	for (var i = 0; i < args.length; i++)
		if (args[i] != null)
			result += args[i];
	return result;
$$
LANGUAGE pljs IMMUTABLE STRICT;
SELECT concat_strings('A', 'B', NULL, 'C');
NOTICE:  Unknown type: 1009
ERROR:  execution error
DETAIL:  TypeError: cannot read property 'length' of null
    at concat_strings (<function>:4)

CREATE FUNCTION return_void() RETURNS void AS $$ $$ LANGUAGE pljs;
SELECT return_void();
 return_void 
-------------
 
(1 row)

CREATE FUNCTION return_null() RETURNS text AS $$ return null; $$ LANGUAGE pljs;
SELECT r, r IS NULL AS isnull FROM return_null() AS r;
  r   | isnull 
------+--------
 null | f
(1 row)

-- TYPE CONVERTIONS
CREATE FUNCTION int2_to_int4(x int2) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT int2_to_int4(24::int2);
 int2_to_int4 
--------------
           24
(1 row)

CREATE FUNCTION int4_to_int2(x int4) RETURNS int2 AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_int2(42);
 int4_to_int2 
--------------
           42
(1 row)

CREATE FUNCTION int4_to_int8(x int4) RETURNS int8 AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_int8(48);
 int4_to_int8 
--------------
           48
(1 row)

CREATE FUNCTION int8_to_int4(x int8) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT int8_to_int4(84);
 int8_to_int4 
--------------
           84
(1 row)

CREATE FUNCTION float8_to_numeric(x float8) RETURNS numeric AS $$ return x; $$ LANGUAGE pljs;
SELECT float8_to_numeric(1.5);
 float8_to_numeric 
-------------------
               1.5
(1 row)

CREATE FUNCTION numeric_to_int8(x numeric) RETURNS int8 AS $$ return x; $$ LANGUAGE pljs;
SELECT numeric_to_int8(1234.56);
 numeric_to_int8 
-----------------
            1234
(1 row)

CREATE FUNCTION int4_to_text(x int4) RETURNS text AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_text(123);
 int4_to_text 
--------------
 123
(1 row)

CREATE FUNCTION text_to_int4(x text) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT text_to_int4('123');
 text_to_int4 
--------------
          123
(1 row)

SELECT text_to_int4('abc'); -- error
 text_to_int4 
--------------
            0
(1 row)

CREATE FUNCTION int4array_to_textarray(x int4[]) RETURNS text[] AS $$ return x; $$ LANGUAGE pljs;
SELECT int4array_to_textarray(ARRAY[123, 456]::int4[]);
NOTICE:  Unknown type: 1007
NOTICE:  Unknown type: 1009
 int4array_to_textarray 
------------------------
 
(1 row)

CREATE FUNCTION textarray_to_int4array(x text[]) RETURNS int4[] AS $$ return x; $$ LANGUAGE pljs;
SELECT textarray_to_int4array(ARRAY['123', '456']::text[]);
NOTICE:  Unknown type: 1009
NOTICE:  Unknown type: 1007
 textarray_to_int4array 
------------------------
 
(1 row)

