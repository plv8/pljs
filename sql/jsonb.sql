CREATE FUNCTION get_key(key text, json_data jsonb) RETURNS jsonb
LANGUAGE pljs IMMUTABLE STRICT
AS $$
  var val = json_data[key];
  var ret = {};
  ret[key] = val;
  return JSON.stringify(ret);
$$;

CREATE TABLE jsonbonly (
    data jsonb
);

COPY jsonbonly (data) FROM stdin;
{"ok": true}
\.

-- Call twice to test the function cache.
SELECT get_key('ok', data) FROM jsonbonly;
SELECT get_key('ok', data) FROM jsonbonly;
