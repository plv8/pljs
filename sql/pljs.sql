-- CREATE FUNCTION
CREATE FUNCTION pljs_test(keys text[], vals text[]) RETURNS text AS
$$
	var o = {};
	for (var i = 0; i < keys.length; i++)
		o[keys[i]] = vals[i];
	return JSON.stringify(o);
$$
LANGUAGE pljs IMMUTABLE STRICT;
SELECT pljs_test(ARRAY['name', 'age'], ARRAY['Tom', '29']);

CREATE FUNCTION unnamed_args(text[], text[]) RETURNS text[] AS
$$
	var array1 = arguments[0];
	var array2 = $2;
	return array1.concat(array2);
$$
LANGUAGE pljs IMMUTABLE STRICT;
SELECT unnamed_args(ARRAY['A', 'B'], ARRAY['C', 'D']);

CREATE FUNCTION concat_strings(VARIADIC args text[]) RETURNS text AS
$$
	var result = "";
	for (var i = 0; i < args.length; i++)
		if (args[i] != null)
			result += args[i];
	return result;
$$
LANGUAGE pljs IMMUTABLE STRICT;
SELECT concat_strings('A', 'B', NULL, 'C');

CREATE FUNCTION return_void() RETURNS void AS $$ $$ LANGUAGE pljs;
SELECT return_void();

CREATE FUNCTION return_null() RETURNS text AS $$ return null; $$ LANGUAGE pljs;
SELECT r, r IS NULL AS isnull FROM return_null() AS r;

-- TYPE CONVERTIONS
CREATE FUNCTION int2_to_int4(x int2) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT int2_to_int4(24::int2);
CREATE FUNCTION int4_to_int2(x int4) RETURNS int2 AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_int2(42);
CREATE FUNCTION int4_to_int8(x int4) RETURNS int8 AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_int8(48);
CREATE FUNCTION int8_to_int4(x int8) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT int8_to_int4(84);
CREATE FUNCTION float8_to_numeric(x float8) RETURNS numeric AS $$ return x; $$ LANGUAGE pljs;
SELECT float8_to_numeric(1.5);
CREATE FUNCTION numeric_to_int8(x numeric) RETURNS int8 AS $$ return x; $$ LANGUAGE pljs;
SELECT numeric_to_int8(1234.56);
CREATE FUNCTION int4_to_text(x int4) RETURNS text AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_text(123);
CREATE FUNCTION text_to_int4(x text) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT text_to_int4('123');
SELECT text_to_int4('abc'); -- error
CREATE FUNCTION int4array_to_textarray(x int4[]) RETURNS text[] AS $$ return x; $$ LANGUAGE pljs;
SELECT int4array_to_textarray(ARRAY[123, 456]::int4[]);
CREATE FUNCTION textarray_to_int4array(x text[]) RETURNS int4[] AS $$ return x; $$ LANGUAGE pljs;
SELECT textarray_to_int4array(ARRAY['123', '456']::text[]);
