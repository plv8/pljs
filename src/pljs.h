#pragma once

#include "postgres.h"

#include "access/heapam.h"
#include "access/htup.h"
#include "access/tupdesc.h"
#include "executor/spi.h"
#include "fmgr.h"
#include "nodes/params.h"
#include "parser/parse_node.h"

#include "deps/quickjs/quickjs-libc.h"
#include "deps/quickjs/quickjs.h"
#include "utils/palloc.h"

#define STORAGE_HASH_LEN 32
#ifndef PLJS_VERSION
#define PLJS_VERSION "unknown"
#endif

// pljs current runtime configuration.
typedef struct pljs_configuration {
  size_t memory_limit;
  char *start_proc;
  int execution_timeout;
} pljs_configuration;

// Global #pljs_configuration configuration.
extern pljs_configuration configuration;

// quickjs runtime.
extern JSRuntime *rt;

// Cpntext cache value definition.
typedef struct pljs_context_cache_value {
  Oid user_id;
  JSContext *ctx;
  MemoryContext function_memory_context;
  HTAB *function_hash_table;
} pljs_context_cache_value;

// Function cache value defition.
typedef struct pljs_function_cache_value {
  Oid fn_oid;
  JSValue fn;
  JSContext *ctx;
  bool trigger;
  Oid user_id;
  int nargs;
  char proname[NAMEDATALEN];
  Oid argtypes[FUNC_MAX_ARGS];
  char *prosrc;
} pljs_function_cache_value;

typedef struct pljs_param_state {
  Oid *param_types;
  int nparams;
  MemoryContext memory_context;
} pljs_param_state;

// Expanded type definitions for pljs.
typedef struct pljs_type {
  Oid typid;
  Oid ioparam;
  int16 len;
  bool byval;
  char align;
  char category;
  bool is_composite;
  FmgrInfo fn_input;
  FmgrInfo fn_output;
} pljs_type;

// Plan for prepared statements.
typedef struct pljs_plan {
  SPIPlanPtr plan;
  pljs_param_state *parstate;
} pljs_plan;

// Context and information for the function to be called.
typedef struct pljs_func {
  Oid fn_oid; // function's OID

  char proname[NAMEDATALEN]; // the function name
  char *prosrc;              // a copy of its source

  TransactionId fn_xmin;
  ItemPointerData fn_tid;
  Oid user_id; // the user id

  bool trigger;
  int nargs;                   // the number of arguments
  bool is_srf;                 // are we a set returning function?
  Oid rettype;                 // the return type
  Oid argtypes[FUNC_MAX_ARGS]; // the types of the argument passed
} pljs_func;

typedef struct pljs_context {
  JSContext *ctx;
  JSValue js_function; // the function itself

  char *arguments[FUNC_MAX_ARGS];
  MemoryContext memory_context;
  pljs_func *function;
} pljs_context;

// Functions in pljs.c
JSValue js_throw(JSContext *, const char *);
void _PG_init(void);
void pljs_guc_init(void);
void pljs_cache_init(void);
void pljs_setup_namespace(JSContext *);
JSValue pljs_compile_function(pljs_context *context, bool is_trigger);

// Functions in cache.c
void pljs_cache_context_add(Oid, JSContext *);
void pljs_cache_context_remove(Oid);
pljs_function_cache_value *pljs_cache_function_find(Oid user_id, Oid fn_oid);
void pljs_cache_function_add(pljs_context *context);
void pljs_cache_function_remove(Oid user_id, Oid fn_oid);
pljs_context_cache_value *pljs_cache_context_find(Oid user_id);

void pljs_function_cache_to_context(pljs_context *,
                                    pljs_function_cache_value *);
void pljs_context_to_function_cache(pljs_function_cache_value *function_entry,
                                    pljs_context *context);

// Functions in type.c
uint32_t js_array_length(JSContext *, JSValue);
void pljs_type_fill(pljs_type *, Oid);
JSValue pljs_datum_to_jsvalue(Datum arg, Oid type, JSContext *ctx);
JSValue pljs_datum_to_array(Datum arg, pljs_type *type, JSContext *ctx);
JSValue pljs_datum_to_object(Datum arg, pljs_type *type, JSContext *ctx);

Datum pljs_jsvalue_to_array(JSValue, pljs_type *, JSContext *,
                            FunctionCallInfo);
Datum pljs_jsvalue_to_datum(JSValue, Oid, JSContext *, FunctionCallInfo,
                            bool *);
Datum pljs_jsvalue_to_record(JSValue val, pljs_type *type, JSContext *ctx,
                             bool *is_null, TupleDesc);
JSValue values_to_array(JSContext *, JSValue *, int, int);
JSValue tuple_to_jsvalue(JSContext *ctx, TupleDesc, HeapTuple);
JSValue spi_result_to_jsvalue(JSContext *, int);

void pljs_variable_param_setup(ParseState *, void *);
ParamListInfo pljs_setup_variable_paramlist(pljs_param_state *, Datum *,
                                            char *);
