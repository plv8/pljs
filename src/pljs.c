#include "deps/quickjs/quickjs.h"
#include "postgres.h"

#include "access/xlog_internal.h"
#include "catalog/pg_proc.h"
#include "catalog/pg_type_d.h"
#include "commands/trigger.h"
#include "common/hashfn.h"
#include "executor/spi.h"
#include "funcapi.h"
#include "miscadmin.h"
#include "nodes/parsenodes.h"
#include "utils/builtins.h"
#include "utils/guc.h"
#include "utils/jsonb.h"
#include "utils/palloc.h"
#include "utils/syscache.h"

#include "pljs.h"

Datum pljs_call_handler(PG_FUNCTION_ARGS);
Datum pljs_call_validator(PG_FUNCTION_ARGS);
Datum pljs_inline_handler(PG_FUNCTION_ARGS);

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(pljs_call_handler);
PG_FUNCTION_INFO_V1(pljs_call_validator);
PG_FUNCTION_INFO_V1(pljs_inline_handler);

static char *dump_error(JSContext *ctx) {
  JSValue exception_val, val;
  const char *stack;
  const char *str;
  bool is_error;
  char *ret = NULL;
  size_t s1, s2;

  exception_val = JS_GetException(ctx);

  /* In the case of OOM, a null exception is thrown. */
  if (JS_IsNull(exception_val)) {
    char *oom = palloc(14);
    strcpy(oom, "out of memory");

    JS_FreeValue(ctx, exception_val);

    return oom;
  }

  is_error = JS_IsError(ctx, exception_val);
  str = JS_ToCStringLen(ctx, &s1, exception_val);

  if (!str) {
    elog(NOTICE, "error thrown but no error message");
    return NULL;
  }

  if (!is_error) {
    ret = (char *)palloc((s1 + 8) * sizeof(char));
    sprintf(ret, "Throw:\n%s", str);
  } else {
    val = JS_GetPropertyStr(ctx, exception_val, "stack");

    if (!JS_IsUndefined(val)) {
      stack = JS_ToCStringLen(ctx, &s2, val);

      ret = (char *)palloc((s1 + s2 + 2) * sizeof(char));
      sprintf(ret, "%s\n%s", str, stack);
      JS_FreeCString(ctx, stack);
    }

    JS_FreeValue(ctx, val);
  }

  JS_FreeCString(ctx, str);
  JS_FreeValue(ctx, exception_val);

  return ret;
}

JSRuntime *rt = NULL;
static uint64_t os_pending_signals = 0;

pljs_configuration configuration = {0};

static void signal_handler(int sig_num) {
  os_pending_signals |= ((uint64_t)1 << sig_num);
}

static int interrupt_handler(JSRuntime *rt, void *opaque) {
  return (os_pending_signals >> SIGINT) & 1;
}

// initialization function
void _PG_init(void) {
  signal(SIGINT, signal_handler);
  signal(SIGTERM, signal_handler);
  signal(SIGABRT, signal_handler);

  // initialize cache
  pljs_cache_init();

  // initialize the GUCs
  pljs_guc_init();

  // set up the quickjs runtime
  rt = JS_NewRuntime();

  // set up a memory limit if it exists
  if (configuration.memory_limit) {
    JS_SetMemoryLimit(rt, configuration.memory_limit * 1024 * 1024);
  }
}

// sets up the configuration of the extension
void pljs_guc_init() {
#ifdef EXECUTION_TIMEOUT
  DefineCustomIntVariable(
      "pljs.execution_timeout", gettext_noop("Javascriot execution timeout."),
      gettext_noop(
          "The default value is 300 seconds."
          "This allows you to override the default execution timeout."),
      &configuration.execution_timeout, 300, 1, 65536, PGC_USERSET, 0, NULL,
      NULL, NULL);
#endif

  DefineCustomIntVariable("pljs.memory_limit",
                          gettext_noop("Runtime limit in MBytes"),
                          gettext_noop("The default value is 256 MB"),
                          (int *)&configuration.memory_limit, 256, 256, 3096,
                          PGC_SUSET, 0, NULL, NULL, NULL);

  DefineCustomStringVariable(
      "pljs.start_proc",
      gettext_noop("PLJS function to run once when PLJS is first used."), NULL,
      &configuration.start_proc, NULL, PGC_USERSET, 0, NULL, NULL, NULL);
}

Datum pljs_call_handler(PG_FUNCTION_ARGS) {
  Oid fn_oid = fcinfo->flinfo->fn_oid;
  HeapTuple proctuple;
  Form_pg_proc pg_proc_entry = NULL;
  const char *sourcecode;
  char **arguments;
  Oid *argtypes = NULL;
  char *argmodes;
  int nargs = 0;
  Datum prosrcdatum;
  bool isnull;
  JSContext *ctx;
  Datum retval;
  JSValue func;
  bool nonatomic = fcinfo->context && IsA(fcinfo->context, CallContext) &&
                   !castNode(CallContext, fcinfo->context)->atomic;
  ;

  proctuple = SearchSysCache(PROCOID, ObjectIdGetDatum(fn_oid), 0, 0, 0);

  if (!HeapTupleIsValid(proctuple)) {
    elog(ERROR, "cache lookup failed for function %u", fn_oid);
  }

  prosrcdatum =
      SysCacheGetAttr(PROCOID, proctuple, Anum_pg_proc_prosrc, &isnull);

  if (isnull) {
    elog(ERROR, "null prosrc");
  }

  pg_proc_entry = (Form_pg_proc)GETSTRUCT(proctuple);
  nargs = get_func_arg_info(proctuple, &argtypes, &arguments, &argmodes);

  pljs_cache_value *entry = NULL; // pljs_cache_function_find(fn_oid);
  if (entry) {
    // elog(NOTICE, "function found");
    ctx = entry->ctx;
    func = entry->fn;
  } else {
    // elog(NOTICE, "no function found");
    //  check to see if a context exists in the cache for this user
    entry = pljs_cache_context_find(GetUserId());

    if (entry) {
      ctx = entry->ctx;
    } else {
      // create a new execution context.
      ctx = JS_NewContext(rt);

      // set up the namespace, globals and functions available inside the
      // context.
      pljs_setup_namespace(ctx);

      // save the context
      pljs_cache_context_add(GetUserId(), ctx);
    }

    sourcecode = DatumGetCString(DirectFunctionCall1(textout, prosrcdatum));

    func = pljs_compile_function(ctx, NameStr(pg_proc_entry->proname),
                                 sourcecode, nargs, arguments);
    if (JS_IsUndefined(func)) {
      // JS_FreeContext(ctx);
      PG_RETURN_VOID();
    }

    // create the cache entry for the function.
    // pljs_cache_function_add(fn_oid, ctx, func);
  }

  if (SPI_connect_ext(nonatomic ? SPI_OPT_NONATOMIC : 0) != SPI_OK_CONNECT) {
    elog(ERROR, "could not connect to spi manager");
  }

  // do the logic
  retval = call_function(fcinfo, ctx, func, nargs, argtypes,
                         pg_proc_entry->prorettype);

  // do not destroy the context, it will get reused
  // JS_FreeContext(ctx);

  ReleaseSysCache(proctuple);

  SPI_finish();

  return retval;
}

Datum pljs_inline_handler(PG_FUNCTION_ARGS) {
  pljs_cache_value *entry = pljs_cache_context_find(GetUserId());

  InlineCodeBlock *code_block =
      (InlineCodeBlock *)DatumGetPointer(PG_GETARG_DATUM(0));
  char *sourcecode = code_block->source_text;

  HeapTuple proctuple;
  Form_pg_proc pg_proc_entry = NULL;
  JSContext *ctx = NULL;
  bool nonatomic = fcinfo->context && IsA(fcinfo->context, CallContext) &&
                   !castNode(CallContext, fcinfo->context)->atomic;

  if (entry) {
    ctx = entry->ctx;
  } else {
    // create a new execution context.
    ctx = JS_NewContext(rt);

    // set up the namespace, globals and functions available inside the
    // context.
    pljs_setup_namespace(ctx);

    // save the context
    pljs_cache_context_add(GetUserId(), ctx);
  }

  if (SPI_connect_ext(nonatomic ? SPI_OPT_NONATOMIC : 0) != SPI_OK_CONNECT) {
    elog(ERROR, "could not connect to spi manager");
  }

  // do the logic
  pljs_call_anonymous_function(ctx, sourcecode);

  SPI_finish();

  PG_RETURN_VOID();
}

Datum pljs_call_validator(PG_FUNCTION_ARGS) {
  Oid fn_oid = fcinfo->flinfo->fn_oid;
  HeapTuple proctuple;
  const char *sourcecode;
  Datum prosrcdatum;
  bool isnull;
  JSContext *ctx;

  if (fcinfo->flinfo->fn_extra) {
    elog(NOTICE, "fn_extra on validate");
  }
  proctuple = SearchSysCache(PROCOID, ObjectIdGetDatum(fn_oid), 0, 0, 0);

  if (!HeapTupleIsValid(proctuple)) {
    elog(ERROR, "cache lookup failed for function %u", fn_oid);
  }

  prosrcdatum =
      SysCacheGetAttr(PROCOID, proctuple, Anum_pg_proc_prosrc, &isnull);

  if (isnull) {
    elog(ERROR, "null prosrc");
  }

  sourcecode = TextDatumGetCString(prosrcdatum);

  ctx = JS_NewContext(rt);

  JSValue val = JS_Eval(ctx, sourcecode, strlen(sourcecode), "<function>",
                        JS_EVAL_FLAG_COMPILE_ONLY);

  if (JS_IsException(val)) {
    ereport(ERROR,
            (errmsg("execution error"), errdetail("%s", dump_error(ctx))));
  }

  // call validator can release the context
  JS_FreeContext(ctx);

  ReleaseSysCache(proctuple);

  PG_RETURN_VOID();
}

JSValue pljs_compile_function(JSContext *ctx, char *name, const char *source,
                              int nargs, char *arguments[]) {
  StringInfoData src;
  int i;

  initStringInfo(&src);

  // generate the function as javascript with all of its arguments
  appendStringInfo(&src, "function %s (", name);

  for (i = 0; i < nargs; i++) {
    // commas between arguments
    if (i > 0) {
      appendStringInfoChar(&src, ',');
    }

    // if this is a named argument, append it
    if (arguments && arguments[i]) {
      appendStringInfoString(&src, arguments[i]);
    } else {
      // otherwise append it as an unnamed argument with a number
      appendStringInfo(&src, "$%d", i + 1);
    }
  }

  // append the other postgres-specific variables as well
  if (nargs) {
    appendStringInfo(&src, ", ");
  }

  appendStringInfo(&src, "NEW, OLD, TG_NAME, TG_WHEN, TG_LEVEL, TG_OP, "
                         "TG_RELID, TG_TABLE_NAME, TG_TABLE_SCHEMA, TG_ARGV");

  appendStringInfo(&src, ") {\n%s\n}\n %s;\n", source, name);

  JSValue val = JS_Eval(ctx, src.data, strlen(src.data), "<function>", 0);

  if (!JS_IsException(val)) {
    pfree(src.data);

    return val;
  } else {
    ereport(ERROR,
            (errmsg("execution error"), errdetail("%s", dump_error(ctx))));

    return JS_UNDEFINED;
  }
}

void pljs_call_anonymous_function(JSContext *ctx, const char *source) {
  StringInfoData src;

  initStringInfo(&src);

  // generate the function as javascript with all of its arguments
  appendStringInfo(&src, "(function () {\n%s\n})();", source);

  JS_SetInterruptHandler(JS_GetRuntime(ctx), interrupt_handler, NULL);
  os_pending_signals &= ~((uint64_t)1 << SIGINT);

  JSValue val = JS_Eval(ctx, src.data, strlen(src.data), "<function>", 0);

  if (!JS_IsException(val)) {
    pfree(src.data);
  } else {

    ereport(ERROR,
            (errmsg("execution error"), errdetail("%s", dump_error(ctx))));
  }
}

static Datum call_function(FunctionCallInfo fcinfo, JSContext *ctx,
                           JSValue func, int nargs, Oid *argtypes,
                           Oid rettype) {
  JSValueConst *argv = (JSValueConst *)palloc(sizeof(JSValueConst) * nargs);

  MemoryContext old_context = CurrentMemoryContext;
  MemoryContext execution_context = AllocSetContextCreate(
      CurrentMemoryContext, "PLJS Memory Context", ALLOCSET_SMALL_SIZES);

  for (int i = 0; i < nargs; i++) {
    if (fcinfo->args[i].isnull == 1) {
      argv[i] = JS_NULL;
    } else {
      argv[i] = pljs_datum_to_jsvalue(fcinfo->args[i].value, argtypes[i], ctx);
    }
  }

  JS_SetInterruptHandler(JS_GetRuntime(ctx), interrupt_handler, NULL);
  os_pending_signals &= ~((uint64_t)1 << SIGINT);

  JSValue ret = JS_Call(ctx, func, JS_UNDEFINED, nargs, argv);

  if (JS_IsException(ret)) {
    ereport(ERROR,
            (errmsg("execution error"), errdetail("%s", dump_error(ctx))));

    JS_FreeValue(ctx, ret);

    CurrentMemoryContext = old_context;
    PG_RETURN_VOID();
  } else {
    Datum d = pljs_jsvalue_to_datum(ret, rettype, ctx, fcinfo, NULL);
    JS_FreeValue(ctx, ret);

    CurrentMemoryContext = old_context;
    return d;
  }
}

JSValue js_throw(JSContext *ctx, const char *message) {
  JSValue error = JS_NewError(ctx);
  JSValue message_value = JS_NewString(ctx, message);
  JS_SetPropertyStr(ctx, error, "message", message_value);

  return JS_Throw(ctx, error);
}
