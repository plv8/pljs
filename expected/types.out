CREATE FUNCTION return_void() RETURNS void AS $$ $$ LANGUAGE pljs;
SELECT return_void();
 return_void 
-------------
 
(1 row)

CREATE FUNCTION return_null() RETURNS text AS $$ return null; $$ LANGUAGE pljs;
SELECT r, r IS NULL AS isnull FROM return_null() AS r;
 r | isnull 
---+--------
   | t
(1 row)

-- TYPE CONVERTIONS
CREATE FUNCTION int2_to_int4(x int2) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT int2_to_int4(24::int2);
 int2_to_int4 
--------------
           24
(1 row)

CREATE FUNCTION int4_to_int2(x int4) RETURNS int2 AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_int2(42);
 int4_to_int2 
--------------
           42
(1 row)

CREATE FUNCTION int4_to_int8(x int4) RETURNS int8 AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_int8(48);
 int4_to_int8 
--------------
           48
(1 row)

CREATE FUNCTION int8_to_int4(x int8) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT int8_to_int4(84);
 int8_to_int4 
--------------
           84
(1 row)

CREATE FUNCTION float8_to_numeric(x float8) RETURNS numeric AS $$ return x; $$ LANGUAGE pljs;
SELECT float8_to_numeric(1.5);
 float8_to_numeric 
-------------------
               1.5
(1 row)

CREATE FUNCTION numeric_to_int8(x numeric) RETURNS int8 AS $$ return x; $$ LANGUAGE pljs;
SELECT numeric_to_int8(1234.56);
 numeric_to_int8 
-----------------
            1234
(1 row)

CREATE FUNCTION int4_to_text(x int4) RETURNS text AS $$ return x; $$ LANGUAGE pljs;
SELECT int4_to_text(123);
 int4_to_text 
--------------
 123
(1 row)

CREATE FUNCTION text_to_int4(x text) RETURNS int4 AS $$ return x; $$ LANGUAGE pljs;
SELECT text_to_int4('123');
 text_to_int4 
--------------
          123
(1 row)

SELECT text_to_int4('abc');
 text_to_int4 
--------------
            0
(1 row)

-- ARRAYS
CREATE FUNCTION return_array() RETURNS TEXT[] AS $$ return ["foo", "bar"]; $$LANGUAGE pljs;
SELECT return_array();
 return_array 
--------------
 {foo,bar}
(1 row)

